README.md
# test-task-getshop

Приложение реализует кеш-сервер для погоды. Сервер позволяет узнать погоду по городу, id города (может быть найден тут: http://bulk.openweathermap.org/sample/), zip-коду или координатам.

### Запросы к серверу

По городу

    /city?q={Название города}
    /city?q=London

По id

    /cityid/?id={Идентификатор}
    /cityid/?id=2172797
    
По зипкоду

    /zipcode/?zip={zip-код}
    /zipcode/?zip=94040,us

По координатам 

    /coords/?lat={Широта}&lon={Долгота}
    /coords/?lat=35&lon=139

### Настройки сервера

Ключ для OpenWheaterMap передается посредством аргумента командной строки

    stack run -- --api {ключ}

Все остальные параметры настраиваются посредством конфигурационного файла config/congif.yaml

#### Параметры

port: порт, на котором будет запущен сервер

    port: 8004

timeError: опционально, допустимая погрешность для времени, еденица измерения -- секунды, значение по умолчанию -- 1200 (20 минут)

    timeError: 3600

rangeError: опционально, допустимая погрешность для локации, еденица измерения -- километры, значение по умолчанию -- 30

    rangeError: 100

updateTime: опционально, промежуток времени между автоматическими обновлениями, еденица измерения -- секунды, значение по умолчанию -- 1800 (30 минут)

    udpateTime: 3600

locations: список локаций, для которых будет производится автоматическое обновление данных

##### Формат данных локаций

typ: показывает, по каким данным будут производится запросы

    typ = "city" | "cityId" | "zipCode" | "coords"

val: для поиска по городу, id и зипкоду:
- "city": Название города
- "cityId": Идентификатор
- "zipCode": zip-код локации

        - typ: city
          val: London
        - typ: cityId
          val: "2172797"
        - typ: zipCode
          val: "94040,us"

lat/lon: для поиска по координатам. Задает широту и долготу соответственно

        - typ: coords
          lat: "35"
          lon: "139"

### Внутреннее устройство

Для хранения данных используется Redis. Всего используется 4 хранилища:
- Temperature_Ordered_Set : Ordered Set
- Cities_Coords_Map : HashMap
- CitiesIDs_Coords_Map : HashMap
- zipcode_Coords_Map : HashMap

Первое используется для хранения JSON'а с погодой по координатам, оставшиеся три задают соответствие  (Описание места -> его координаты). 

Когда происходит первый запрос локации не по координатам сервер честно отправляет запрос на сервер OpenWheatherMap и из полученого JSON'а получает координаты, которые сохраняет в HashMap'е 

Данные в Ordered Set'е хранятся упорядоченные по долготе (longitude)

Для быстрого доcтупа к данным по координатам программа изначально высчитывает две точки, которые находятся на длине растояния погрешности от данного места, если при этом идти строго на запад или восток. Теперь, если взять долготу этих точек, мы можем получить сектор, в котором обязательно будут лежать все точки, которые находятся на расстоянии меньше расстояния погрешности. 

![alt Explanation_Picture](explanation.png)

После этого из Set'а берутся все точки в указаном сегменте, и среди них линейно ищется ближайшая.

При каждом обращении к серверам OpenWheatherMap происходит обновление данных в Set'e


ChangeLog.md
# Changelog for test-task-getshop

## Unreleased changes


src/Utility/Config.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}

module Utility.Config (readConfig) where

import           Control.Exception       (catch)
import           Data.Aeson              (FromJSON, (.:))
import           Data.Maybe              (fromJust, isJust)
import           Data.Text               (Text)
import           Data.Yaml               ((.:?))
import qualified Data.Yaml               as Yaml
import           Types.Config            (ConfigFields (..))
import           Types.Wheather          (CoordsStr (CoordsStr),
                                          RequestType (City, CityId, Coords, ZIPCode))
import           Utility.ConfigException (ConfigExc (getExcText),
                                          throwConfigExc)

readConfig :: IO (Either String ConfigFields)
readConfig =
  catch
    ( do
        eObj <- Yaml.decodeFileEither "config/config.yaml"
        obj <- either (throwConfigExc . Yaml.prettyPrintParseException) pure eObj
        timeError <- readValueDef "timeError" obj 1200
        rangeError <- readValueDef "rangeError" obj 30
        port <- readValue "port" obj
        autoUpdLocations <- readPlacesList obj
        sleepTime <- readValueDef "sleepTime" obj 1800
        pure . Right $ ConfigFields {..}
    )
    (\e -> pure . Left . getExcText $ (e :: ConfigExc))

readValue :: FromJSON a => Text -> Yaml.Object -> IO a
readValue field obj = do
  let eValue = Yaml.parseEither (.: field) obj
  either throwConfigExc pure eValue

readValueDef :: FromJSON a => Text -> Yaml.Object -> a -> IO a
readValueDef field obj def = do
  let eValue = Yaml.parseEither (.: field) obj
  either (pure . const def) pure eValue

readPlacesList :: Yaml.Object -> IO [RequestType]
readPlacesList obj = do
  let lst = Yaml.parseEither (.: "locations") obj
  case lst of
    Left err -> throwConfigExc err
    Right lst_ -> do
      let locations = map (Yaml.parseEither parseLocations) lst_
      mapM helper locations
  where
    helper (Left err)  = throwConfigExc err
    helper (Right loc) = getRequestType loc

    getRequestType :: Location -> IO RequestType
    getRequestType Location {..} = do
      case typ of
        "city" -> maybe (throwConfigExc "Bad location") (pure . City) val
        "cityId" -> maybe (throwConfigExc "Bad location") (pure . CityId) val
        "zipCode" -> maybe (throwConfigExc "Bad location") (pure . ZIPCode) val
        "coords" -> if isJust lat && isJust lon then pure . Coords $ CoordsStr (fromJust lat) (fromJust lon) else throwConfigExc "Bad location"
        _ -> throwConfigExc "Bad location"

data Location = Location
  { typ :: String,
    val :: Maybe String,
    lat :: Maybe String,
    lon :: Maybe String
  }

parseLocations :: Yaml.Object -> Yaml.Parser Location
parseLocations v =
  Location
    <$> v .: "typ"
    <*> v .:? "val"
    <*> v .:? "lat"
    <*> v .:? "lon"
